### DISCUSSION TITLE
@change in script tag

### DISCUSSION THREAD

**Discussion by jamesgrubb** [alpinejs/alpine#906] [4. General]
Sorry this is a dumb question but if I'm writing my data ina script tag can I still use @ change

```function data(){
return{
@change...
}

}```
or do I need to use  `this.$refs.fileInput.addEventListener('change',function(){})`
Thanks

**Comment by HugoDF** [comment]
You can use `@change` with x-spread or directly

```html
<div x-data="hello()" x-spread="stuff" @change="foo"></div>
<script>
function hello() {
  return {
    foo(event) { /* handle click */ },
    stuff: {
      ['@change']() { /* handle click */ }
    }
  }
}
</script>
```

**Reply by jamesgrubb** [reply]
Thank you! I'm guessing that @click was meant to be @change as that is what the original question was :)

**Reply by HugoDF** [reply]
whoops my bad I'll update it to `@change` but yeah same applies

**Reply by jamesgrubb** [reply]
Thank you

**Reply by jamesgrubb** [reply]
So something like this:
```
<div x-data="hello()" x-spread="stuff" @click="foo">
  <input type="file" x-ref="fileInput">
  <p x-text="stuff"></p>
</div>
```

```
function hello() {
  return {
    foo(event) { /* handle click */ },
    stuff: {
      ['@change']() { 
        return(this.$refs.fileInput.files[0].name )
      }
    }
  }
}
```

Except that works :(

**Reply by HugoDF** [reply]
Right the wording of your question was confusing, you can do this:
```html
<div x-data="hello()">
  <input type="file" @change="runSomeFunction($event)">
  <p x-text="stuff"></p>
</div>
<script>
function hello() {
  return {
    runSomeFunction($event) {
      this.stuff = $event.target.files[0].name;
    },
    stuff: ''
  }
}
</script>
```

**Reply by jamesgrubb** [reply]
I am very confused most of the time. Thanks again

**Reply by jamesgrubb** [reply]
Im getting image data back but its not being passed to :src?

const reader = new FileReader();
```
function hello() {
  return {
    runSomeFunction($event) {
      reader.onload = async function () {
        console.log(reader.result)
        this.stuff = await reader.result;
      };
      reader.readAsDataURL($event.target.files[0]);
    },
    stuff: ""
  };
}

```

```
<div x-data="hello()">
  <input type="file" @change="runSomeFunction($event)">
  <p x-text="stuff"></p>
  <img :src="stuff" alt="">
</div>
```

**Reply by jamesgrubb** [reply]
with or without the 'await'

**Reply by HugoDF** [reply]
With a regular `function` you're losing the `this` context so `this.stuff` isn't the Alpine.js `this.stuff`, try the following:

```js
reader.onload = async () => {
  console.log(reader.result)
  this.stuff = await reader.result;
};
```

**Comment by jamesgrubb** [comment]
I've made [a pen](https://codepen.io/limitedunlimited/pen/ZEpbYBv). Can't work why this is not working. I'm sure its a simple thing. I'm getting the dataURL logged but does not seem to be passed to the img :src Thanks
```
function hello() {
  return {
    imgsrc: "",
    fileInput($event) {
      let reader = new FileReader();
      reader.addEventListener("load", function (event) {
        console.log(event.target.result);
        this.imgsrc = event.target.result;
      });
      reader.readAsDataURL($event.target.files[0]);
    }
  };
}
```

**Reply by KevinBatdorf** [reply]
Hi

Change this line: `reader.addEventListener("load", (event) => {`

**Reply by jamesgrubb** [reply]
BINGO! Thank You! Would you mind explaining? Im guessing something to do with the 'this' keyword?

**Reply by jamesgrubb** [reply]
I thought you couldn't use this inside an arrow function?

**Reply by KevinBatdorf** [reply]
Yeah it's about scope and `this`. You need to use an arrow function here otherwise `this` will be scoped to the inner function.

Here's a good reference for that: https://leanpub.com/understandinges6/read/#leanpub-auto-arrow-functions
